### PR TITLE
Tweak the code, comments and utilities.

### PR DESCRIPTION
**Pull request by lrhn**
Nothing here should change behavior or API, only
implementation and presentation.

Introduced the `Mutex` class because `Pool` seemed 
overqualified/overcomplicated for the job of being a single-lock mutex.

Added a `JsonConvertible` interface, just to have
an extension method on `Iterable<JsonConvertible>` 
for doing `[for (var v in iterable) v.toJson()]`.
Happened too often, so a good candidate for a
helper function.

Every change here is a suggestion, none of it is
essential. Feel free to ignore any parts that
seem unnecessary.
(The task-URI cache is probably unnecessary.)

### CODE REVIEW

File: samples/dart/bin/util/resource.dart [hunk 1]
```diff
@@ -17,7 +17,16 @@ import 'dart:typed_data';
...
 
+/// Reads a file from the `resources/` directory of this Pub package root
+/// directory.
+///
+/// The [name] is the entire file name, including extension.
+///
+/// Assumes the script is run from the `bin/` directory or another immediate
+/// subdirectory of the package root directory.
 Future<Uint8List> readResource(String name) {
-  final uri = Platform.script.resolve(path.join('..', 'resources', name));
```
Comment (lrhn) [inline samples/dart/bin/util/resource.dart:21]: This code would (maybe) not work on Windows.
The `path.join` would introduce backslashes, and `Uri.resolve` would not treat them as path separators.
Maybe `File.fromUri` would retain the `\`s, and then it worked anyway, but along the way, the URI was not a valid `file:` URI for the file.

(Never mix file paths and URIs. Always go between them explicitly using `Directory.fromUri(fileUro)` or `Uri.file(path)`, and only use `package:path` on file paths.)

File: pkgs/google_generative_ai/lib/google_generative_ai.dart [hunk 2]
```diff
@@ -12,12 +12,13 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/// Google generative AI SDK API bindings for Dart.
```
Comment (devoncarew) [inline pkgs/google_generative_ai/lib/google_generative_ai.dart:15]: As a drive-by-comment, if updating this text, can you also update the package description in the pubspec?

File: pkgs/google_generative_ai/pubspec.yaml [hunk 3]
```diff
@@ -11,7 +11,7 @@ environment:
 
 dependencies:
   http: ^1.1.0
-  pool: ^1.5.0
+  meta: ^1.11.0
```
Comment (devoncarew) [inline pkgs/google_generative_ai/pubspec.yaml:14]: This should address the CI failure:

```suggestion
  meta: ^1.10.0
```

File: pkgs/google_generative_ai/lib/google_generative_ai.dart (hunk 2 above)
Comment (lrhn) [reply pkgs/google_generative_ai/lib/google_generative_ai.dart:15]: Done, PTAL

File: pkgs/google_generative_ai/lib/src/chat.dart [hunk 4]
```diff
@@ -88,61 +88,72 @@ final class ChatSession {
   /// be reflected in the history sent for this message.
   ///
   /// Waits to read the entire streamed response before recording the message
-  /// and response and allow pending messages to be sent.
-  Stream<GenerateContentResponse> sendMessageStream(Content message) async* {
-    final resource = await _pool.request();
```
[the remark this answers is not in the document]
Comment (natebosch) [reply pkgs/google_generative_ai/lib/src/chat.dart:93]: I think buffering it in another StreamController makes sense. This will also let us eagerly add it to the history even if the returned stream doesn't have a listener.
If we want to avoid some of the overhead of buffering in two places we could add to history and return a stream that can read from history when it gets a listener.

File: pkgs/google_generative_ai/lib/src/utils/json.dart [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+// limitations under the License.
+
+/// Utilities for working with classes that can be converted to JSON.
+library;
+
+/// General interface that a class with a [toJson] method can implement.
+///
+/// Allows generalized helper functionality that works on different,
+/// otherwise unrelated, classes.
+abstract interface class JsonConvertible {
```
Comment (natebosch) [inline pkgs/google_generative_ai/lib/src/utils/json.dart:22]: I was hoping to avoid exposing serialization details in the package public API. It was too noisy to drop a public `toJson` in my earlier versions, but after changing to a deep `toJson()` it's no longer relying on the behavior of that specific method in `jsonEncode`. I might look at pulling the serialization out into an extension or a private method.

Comment (lrhn) [reply pkgs/google_generative_ai/lib/src/utils/json.dart:22]: ACK.
It really is an internal detail, only existing to allow the extension method on `List<JsonConvertible>`.

If we could make it a private class, `_JsonConvertible`, then it wouldn't leak, but every immediate subclass then needs to be declared in the same library.

File: pkgs/google_generative_ai/lib/src/chat.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/google_generative_ai/lib/src/chat.dart:98]: I don't think what we store in the history, and what's in the stream events, is necessarily the same thing.
The stream events are the original server events, which can (according to the class APIs, at least) include multiple candidates.
The content stored in the history is aggregated text from the *first* candidate of each event.

[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/google_generative_ai/lib/src/chat.dart:98]: Anyway, I rewrote the function to use a stream controller to cache the events.
It was a remarkably simple rewrite.

**Comment by natebosch** [comment]
Closing since I don't think we plan on landing further changes from this branch.
